### PR TITLE
Enabled TLS encryption for PostgreSQL and PgBouncer

### PR DESCRIPTION
**Pull request by klention**
Enabled TLS encryption for PostgreSQL and PGBouncer (related issue: https://github.com/vitabaks/autobase/issues/361)

Generate a Self Signed OpenSSL certificate.

### CODE REVIEW

File: automation/roles/patroni/tasks/main.yml [hunk 1]
```diff
@@ -427,6 +427,22 @@
         state: directory
         mode: "0700"
 
+    # Generating TLS Certificates
+    - name: Generating TLS Certificates to {{ postgresql_home_dir }}/{{ postgresql_version }}
+      become: true
+      become_user: postgres
+      shell:
```
Comment (vitabaks) [inline automation/roles/patroni/tasks/main.yml:434]:
```suggestion
      ansible.builtin.command:
```
It is recommended to use the command module where it is not necessary to use a shell script

File: automation/roles/patroni/tasks/main.yml [hunk 2]
```diff
@@ -427,6 +427,22 @@
...
+        cmd: openssl req -nodes -new -x509 -days 3650 -keyout server.key -out server.crt -subj '/C=AL/L=City/O=Org/CN=PostgreSQL'
+        chdir: "{{ postgresql_home_dir }}/{{ postgresql_version }}/"
+
```
Comment (vitabaks) [inline automation/roles/patroni/tasks/main.yml:437]:
```suggestion

```

File: automation/roles/pgbouncer/tasks/main.yml [hunk 3]
```diff
@@ -124,6 +124,22 @@
     label: "{{ 'pgbouncer' if idx == 0 else 'pgbouncer-%d' % (idx + 1) }}"
   tags: pgbouncer_logrotate, pgbouncer
 
+# Generating TLS Certificates
+- name: Generating TLS Certificates to {{ pgbouncer_conf_dir }}
+  become: true
+  become_user: postgres
+  shell:
```
Comment (vitabaks) [inline automation/roles/pgbouncer/tasks/main.yml:131]:
```suggestion
  ansible.builtin.command:
```

File: automation/roles/pgbouncer/tasks/main.yml [hunk 4]
```diff
@@ -124,6 +124,22 @@
...
+    cmd: openssl req -nodes -new -x509 -days 3650 -keyout server.key -out server.crt -subj '/C=AL/L=City/O=Org/CN=PostgreSQL'
+    chdir: "{{ pgbouncer_conf_dir }}/"
+
+- name: Changing permissions for {{ pgbouncer_conf_dir }}/server.key
+  file:
```
Comment (vitabaks) [inline automation/roles/pgbouncer/tasks/main.yml:136]:
```suggestion
  ansible.builtin.file:
```

File: automation/roles/pgbouncer/tasks/main.yml [hunk 5]
```diff
@@ -124,6 +124,22 @@
...
+    path: "{{ pgbouncer_conf_dir }}/server.key"
+    state: file
+    owner: postgres
+    group: postgres
+    mode: 0400
```
Comment (vitabaks) [inline automation/roles/pgbouncer/tasks/main.yml:141]:
```suggestion
    mode: "0400"
```

File: automation/roles/patroni/tasks/main.yml [hunk 6]
```diff
@@ -427,6 +427,21 @@
...
+      ansible.builtin.command:
+        cmd: openssl req -nodes -new -x509 -days 3650 -keyout server.key -out server.crt -subj '/C=AL/L=City/O=Org/CN=PostgreSQL'
+        chdir: "{{ postgresql_home_dir }}/{{ postgresql_version }}/"
+    - name: Changing permissions for {{ postgresql_home_dir }}/{{ postgresql_version }}/server.key
+      ansible.builtin.file:
+        path: "{{ postgresql_home_dir }}/{{ postgresql_version }}/server.key"
+        state: file
+        owner: postgres
+        group: postgres
+        mode: "0400"
```
Comment (vitabaks) [inline automation/roles/patroni/tasks/main.yml:443]:
```suggestion
    # Generating TLS Certificates
    - name: Generating TLS Certificates to {{ postgresql_home_dir }}/{{ postgresql_version }}
      become: true
      become_user: postgres
      ansible.builtin.command:
        cmd: openssl req -nodes -new -x509 -days 3650 -keyout server.key -out server.crt -subj '/C=AL/L=City/O=Org/CN=PostgreSQL'
        chdir: "{{ postgresql_home_dir }}/{{ postgresql_version }}/"

    - name: Changing permissions for {{ postgresql_home_dir }}/{{ postgresql_version }}/server.key
      ansible.builtin.file:
        path: "{{ postgresql_home_dir }}/{{ postgresql_version }}/server.key"
        state: file
        owner: postgres
        group: postgres
        mode: "0400"
```

File: automation/roles/patroni/tasks/main.yml [hunk 7]
```diff
@@ -427,6 +427,22 @@
         state: directory
         mode: "0700"
 
+    # Generating TLS Certificates
+    - name: Generating TLS Certificates to {{ postgresql_home_dir }}/{{ postgresql_version }}
+      become: true
+      become_user: postgres
+      ansible.builtin.command:
+        cmd: openssl req -nodes -new -x509 -days 3650 -keyout server.key -out server.crt -subj '/C=AL/L=City/O=Org/CN=PostgreSQL'
+        chdir: "{{ postgresql_home_dir }}/{{ postgresql_version }}/"
```
Comment (vitabaks) [inline automation/roles/patroni/tasks/main.yml:436]: Using the commands for generating certificates using openssl is really not an idempotent approach. Ansible has special modules that allow you to generate certificates

- https://docs.ansible.com/ansible/latest/collections/community/crypto/openssl_privatekey_module.html
- https://docs.ansible.com/ansible/latest/collections/community/crypto/x509_certificate_module.html

Something like this:

```yaml
- name: Generate a private key for TLS
  community.crypto.openssl_privatekey:
    path: "{{ postgresql_home_dir }}/server.key"
    size: 2048
    owner: postgres
    group: postgres
    mode: '0600'

- name: Generate a self-signed certificate for Postgres
  community.crypto.x509_certificate:
    path: "{{ postgresql_home_dir }}/server.crt"
    privatekey_path: "{{ postgresql_home_dir }}/server.key"
    owner: postgres
    group: postgres
    mode: '0644'
    provider: selfsigned
    subject:
      C: "AL"
      L: "City"
      O: "Org"
      CN: "PostgreSQL"
    valid_days: 3650
```

**Comment by vitabaks** [comment]
@klention Thank you for your contribution! Please consider my comments.

File: automation/roles/patroni/tasks/main.yml (hunk 7 above)
Comment (klention) [reply automation/roles/patroni/tasks/main.yml:436]: This is a better approach. Will update it.

**Comment by klention** [comment]
@vitabaks, I updated the scrips as suggested.

**Comment by vitabaks** [comment]
> fatal: [10.172.0.22]: FAILED! => {"changed": false, "msg": "The directory /var/lib/postgresql/17 does not exist or the file is not a directory", "name": "/var/lib/postgresql/17"}

I think we need another path to store certificates.

File: automation/roles/patroni/tasks/main.yml [hunk 8]
```diff
@@ -427,6 +427,25 @@
...
+
+    - name: Generate a self-signed certificate for Postgres
+      community.crypto.x509_certificate:
+        path: "{{ postgresql_home_dir }}/{{ postgresql_version }}/server.crt"
+        privatekey_path: "{{ postgresql_home_dir }}/{{ postgresql_version }}/server.key"
+        owner: postgres
+        group: postgres
+        mode: '0644'
+        provider: selfsigned
+        entrust_not_after: "+1000d"
```
Comment (vitabaks) [inline automation/roles/patroni/tasks/main.yml:447]:
```suggestion
    # Generating TLS Certificates
    - name: Generate a private key for TLS
      become: true
      become_user: postgres
      community.crypto.openssl_privatekey:
        path: "{{ postgresql_conf_dir }}/server.key"
        size: '4096'
        mode: '0400'

    - name: Generate a self-signed certificate for Postgres
      community.crypto.x509_certificate:
        path: "{{ postgresql_conf_dir }}/server.crt"
        privatekey_path: "{{ postgresql_conf_dir }}/server.key"
        owner: postgres
        group: postgres
        mode: '0644'
        provider: selfsigned
        entrust_not_after: "+1000d"
```

File: automation/vars/main.yml [hunk 9]
```diff
@@ -235,6 +235,10 @@ postgresql_parameters:
   - { option: "max_connections", value: "1000" }
   - { option: "superuser_reserved_connections", value: "5" }
   - { option: "password_encryption", value: "{{ postgresql_password_encryption_algorithm }}" }
+  - { option: "ssl", value: "on"}
+  - { option: "ssl_cert_file", value: "{{ postgresql_home_dir }}/{{ postgresql_version }}/server.crt"}
+  - { option: "ssl_key_file", value: "{{ postgresql_home_dir }}/{{ postgresql_version }}/server.key"}
```
Comment (vitabaks) [inline automation/vars/main.yml:240]:
```suggestion
  - { option: "ssl_cert_file", value: "{{ postgresql_conf_dir }}/server.crt"}
  - { option: "ssl_key_file", value: "{{ postgresql_conf_dir }}/server.key"}
```

File: automation/vars/main.yml [hunk 10]
```diff
@@ -366,12 +370,18 @@ pgbouncer_auth_user: true # or 'false' if you want to manage the list of users f
...
+pgbouncer_client_tls_cert_file: "{{ pgbouncer_conf_dir }}/server.crt"
 pgbouncer_client_tls_ca_file: ""
 pgbouncer_client_tls_protocols: "secure" # allowed values: tlsv1.0, tlsv1.1, tlsv1.2, tlsv1.3, all, secure (tlsv1.2,tlsv1.3)
-pgbouncer_client_tls_ciphers: "default" # allowed values: default, secure, fast, normal, all (not recommended)
+pgbouncer_client_tls_ciphers: "secure" # allowed values: default, secure, fast, normal, all (not recommended)
+pgbouncer_server_tls_sslmode: "require"
+pgbouncer_server_tls_protocols: "secure"
+pgbouncer_server_tls_ciphers: "secure"
+pgbouncer_server_tls_cert_file: "{{ pgbouncer_conf_dir }}/server.crt"
+pgbouncer_server_tls_key_file: "{{ pgbouncer_conf_dir }}/server.key"
```
Comment (vitabaks) [inline automation/vars/main.yml:383]: As far as I know, we have to use the Postgres key and certificate here
```suggestion
pgbouncer_server_tls_cert_file: "{{ postgresql_conf_dir }}/server.crt"
pgbouncer_server_tls_key_file: "{{ postgresql_conf_dir }}/server.key"
```

server_tls_cert_file - Certificate for private key. PostgreSQL server can validate it.
server_tls_key_file - Private key for PgBouncer to authenticate against PostgreSQL server.

Source: https://www.pgbouncer.org/config.html

File: automation/vars/main.yml [hunk 11]
```diff
@@ -366,12 +370,18 @@ pgbouncer_auth_user: true # or 'false' if you want to manage the list of users f
 pgbouncer_auth_username: pgbouncer # user who can query the database via the user_search function
 pgbouncer_auth_password: "" # If not defined, a password will be generated automatically during deployment
 pgbouncer_auth_dbname: "postgres"
-pgbouncer_client_tls_sslmode: "disable"
-pgbouncer_client_tls_key_file: ""
-pgbouncer_client_tls_cert_file: ""
+pgbouncer_client_tls_sslmode: "require"
+pgbouncer_client_tls_key_file: "{{ pgbouncer_conf_dir }}/server.key"
+pgbouncer_client_tls_cert_file: "{{ pgbouncer_conf_dir }}/server.crt"
```
Comment (vitabaks) [inline automation/vars/main.yml:375]:
```suggestion
pgbouncer_client_tls_key_file: "{{ postgresql_conf_dir }}/server.key"
pgbouncer_client_tls_cert_file: "{{ postgresql_conf_dir }}/server.crt"
```

We can use the Postgres key and certificate.

File: automation/roles/pgbouncer/tasks/main.yml [hunk 12]
```diff
@@ -124,6 +124,27 @@
...
+      community.crypto.x509_certificate:
+        path: "{{ pgbouncer_conf_dir }}/server.crt"
+        privatekey_path: "{{ pgbouncer_conf_dir }}/server.key"
+        owner: postgres
+        group: postgres
+        mode: '0644'
+        provider: selfsigned
+        entrust_not_after: "+1000d"
+  when: pgbouncer_server_tls_sslmode | default('disable') != 'disable'
+
```
[the remark this answers is not in the document]
Comment (vitabaks) [reply automation/roles/pgbouncer/tasks/main.yml:147]: @klention what do you think?

**Comment by klention** [comment]
> > fatal: [10.172.0.22]: FAILED! => {"changed": false, "msg": "The directory /var/lib/postgresql/17 does not exist or the file is not a directory", "name": "/var/lib/postgresql/17"}
> 
> I think we need another path to store certificates.

This is strange because just one task above [at line 422](https://github.com/vitabaks/autobase/blob/master/automation/roles/patroni/tasks/main.yml#L422) creates the `postgresql_data_dir` directory. But, we can definitely put it somewhere else.

File: automation/vars/main.yml (hunk 10 above)
Comment (klention) [reply automation/vars/main.yml:383]: I might be wrong but I believe this is required if `clientcert=verify-ca` or `clientcert=verify-full` is set on `pg_hba.conf`. That way PostgreSQL can validate the cert and since it is the same it will accept it. Or, having different certs both signed by a CA (which we are not using). Will test your suggestion and get back.

File: automation/vars/main.yml (hunk 11 above)
Comment (klention) [reply automation/vars/main.yml:375]: I set the path of variable pgbouncer_conf_dir since PGBouncer role executes before Patroni role

File: automation/roles/pgbouncer/tasks/main.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (klention) [reply automation/roles/pgbouncer/tasks/main.yml]: Generating one pair of TLS cert sounds good. However, I moved the task for generating the cert/key to the PGBouncer role instead of PostgreSQL role since PGBouncer role executes first. This way PGBouncer tasks will not fail if the cert is not yet created.

[the remark this answers is not in the document]
Comment (vitabaks) [reply automation/roles/pgbouncer/tasks/main.yml]: Maybe we should put this in a separate role? Considering that the TLS key and certificate will be used not only for pgbouncer.

[the remark this answers is not in the document]
Comment (klention) [reply automation/roles/pgbouncer/tasks/main.yml]: Agree. Created a new role named tls_certificate to host TLS generation tasks.
Made it dynamic through vars with parameters like the following example:

    tls_title: "PostgreSQL & PGBouncer" -> Will append this string to the name of the task for debugging purposes.
    tls_privatekey_path: "{{ pgbouncer_conf_dir }}/server.key" -> Full path of the private key location where it will be created.
    tls_cert_path: "{{ pgbouncer_conf_dir }}/server.crt" -> Full path of the certificate location where it will be created.
    tls_owner: "postgres" -> the owner and group for both key/cert

[the remark this answers is not in the document]
Comment (vitabaks) [reply automation/roles/pgbouncer/tasks/main.yml]: I've made some small changes https://github.com/vitabaks/autobase/pull/849/commits/0aed35fbcef6919a78c8dac7aef538ef21bb9466

File: automation/vars/main.yml [hunk 13]
```diff
@@ -305,8 +314,8 @@ postgresql_parameters:
   - { option: "wal_receiver_status_interval", value: "10s" }
   - { option: "idle_in_transaction_session_timeout", value: "10min" }  # reduce this timeout if possible
   - { option: "jit", value: "off" }
-  - { option: "max_worker_processes", value: "24" }
-  - { option: "max_parallel_workers", value: "8" }
+  - { option: "max_worker_processes", value: "{{ [ansible_processor_vcpus | int, 16] | max }}" }
+  - { option: "max_parallel_workers", value: "{{ [(ansible_processor_vcpus | int // 2), 8] | max }}" }
```
Comment (vitabaks) [inline automation/vars/main.yml:318]:
```suggestion
  - { option: "max_worker_processes", value: "{{ [ansible_processor_vcpus | int, 16] | max }}" }
  - { option: "max_parallel_workers", value: "{{ [(ansible_processor_vcpus | int // 2), 8] | max }}" }
```

I propose to make this change in a separate PR

File: automation/vars/main.yml [hunk 14]
```diff
@@ -174,6 +174,12 @@ consul_services:
 #      - { http: "http://{{ inventory_hostname }}:{{ patroni_restapi_port }}/async?lag={{ patroni_maximum_lag_on_replica }}", interval: "2s" }
 #      - { args: ["systemctl", "status", "pgbouncer"], interval: "5s" }
 
+# TLS certificate (for PostgreSQL & PGBouncer)
+tls_cert_generate: true
+tls_cert_valid_days: 3650
```
Comment (klention) [inline automation/vars/main.yml:179]: According to the documentation: The minimum certificate lifetime is 90 days, and maximum is three years.

File: automation/vars/main.yml (hunk 13 above)
Comment (klention) [reply automation/vars/main.yml:319]: Will reset it to its original state and make a new PR as suggested.

**Comment by vitabaks** [comment]
We also need to ensure that the certificate and key are copied to the new cluster node during scaling (add_pgnode.yml). If this is challenging for you, I’m happy to handle it.

**Comment by klention** [comment]
> We also need to ensure that the certificate and key are copied to the new cluster node during scaling (add_pgnode.yml). If this is challenging for you, I’m happy to handle it.

Here is the new [commit](https://github.com/vitabaks/autobase/pull/849/commits/3097f7ae1bd4bc4cffd043d06839dd7cba59a4ea)

**Comment by vitabaks** [comment]
> Here is the new commit

Thanks, I made a small change ([commit](https://github.com/vitabaks/autobase/pull/849/commits/66b35227b33263b9051c37561fbea6f85e4e8a15)) and tested it, everything works!
